File: leetcode/q215/quick_select.cpp
#include <vector>
#include <iostream>
#include <ctime>
#include <cstdlib>

using namespace std;
class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        if (nums.empty() || k <= 0 || k > nums.size())
        {
            return -1;
        }

        int l = 0, r = nums.size() - 1, p;
        while (l <= r)
        {
            p = getPartition(nums, l, r);
            if (nums.size() - k == p)
            {
                return nums[p];
            }
            if (nums.size() - k > p)
            {
                l = p + 1;
            }
            else
            {
                r = p - 1;
            }
        }
        return -1;
    }

private:
    int getPartition(vector<int> &nums, int l, int r)
    {
        srand(time(0));
        int pIndex = l + rand() % (r - l + 1);
        int pValue = nums[pIndex];
        swap(nums, pIndex, nums.size() - 1);
        int leftBound = l;
        int rightBound = r - 1;
        while (leftBound <= rightBound)
        {
            if (nums[leftBound] <= pValue)
            {
                leftBound++;
            }
            else if (nums[rightBound] > pValue)
            {
                rightBound--;
            }
            else
            {
                swap(nums, leftBound++, rightBound--);
            }
        }
        swap(nums, leftBound, r);
        return leftBound;
    }

private:
    void swap(vector<int> &nums, int i, int j)
    {
        if (i == j)
            return;
        nums[i] ^= nums[j];
        nums[j] ^= nums[i];
        nums[i] ^= nums[j];
    }
};

File: leetcode/q215/QuickSelect.java
package leetcode.q215;

public class QuickSelect {

    public int kthLargest(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k <= 0 || nums.length < k) {
            return Integer.MAX_VALUE;
        }

        int left = 0;
        int right = nums.length - 1;
        int pIndex;
        while (left <= right) {
            pIndex = getPartition(nums, left, right);
            if (nums.length - k == pIndex) {
                return nums[pIndex];
            }
            if (nums.length - k > pIndex) {
                left = pIndex + 1;
            } else {
                right = pIndex - 1;
            }
        }
        return Integer.MAX_VALUE;
    }

    private int getPartition(int[] nums, int left, int right) {
        int pIndex = left + (int) (Math.random() * (right - left + 1));
        int pValue = nums[pIndex];
        swap(nums, pIndex, right);
        int leftBound = left;
        int rightBound = right - 1;
        while (leftBound <= rightBound) {
            if (nums[leftBound] <= pValue) {
                leftBound++;
            } else if (nums[rightBound] > pValue) {
                rightBound--;
            } else {
                swap(nums, leftBound++, rightBound--);
            }
        }
        swap(nums, leftBound, right);
        return leftBound;
    }

    private void swap(int[] array, int i, int j) {
        int tmp = array[i];
        array[i] = array[j];
        array[j] = tmp;
    }

    public static void main(String[] args) {
        QuickSelect s = new QuickSelect();
        int[] nums = {3, 2, 1, 5, 6, 4};
        System.out.println(s.kthLargest(nums, 2));
    }
}